Python 3 support

apoikos opened the pull request:
~This is still work-in-progress expect frequent updates/force-pushes~

With this PR all tests pass and I'm able to setup a cluster and perform rudimentary operations without issues. It probably fixes everything that can be detected by static analysis tools and takes care of a lot of runtime issues, but we need comprehensive testing to make sure we don't get any runtime errors, especially because of string vs bytes mismatches. A good idea here would be to go through *all* of the code and add type annotations, which could be used to detect more issues.

Because of its size, this PR probably cannot be reviewed thoroughly. OTOH, it's an atomic conversion to Python 3 that cannot be performed in multiple merges without breaking stuff. Still I tried to make individual commits as self-contained as possible.

dannyman commented on lib/netutils.py:
```diff
@@ -61,12 +60,6 @@
 _STRUCT_UCRED = "iII"
 _STRUCT_UCRED_SIZE = struct.calcsize(_STRUCT_UCRED)
 
-# Workaround a bug in some linux distributions that don't define SO_PEERCRED
```
I can confirm that in Ubuntu 14.04, 16.04 and 18.04, socket(7) refers to `SO_PEERCRED` so I don't see a need to retain this workaround from that perspective.

dannyman commented on test/py/ganeti.utils.log_unittest.py:
```diff
@@ -141,7 +146,10 @@ def testConsole(self):
         self.assertTrue("Test message ERROR" in consout)
 
-  class _FailingFile(file):
+    temp_file.close()
```
The diff gets confusing but these two lines are cleanup added to `testConsole()` and then `failingFile()` gets re-declared with the revised argument.

iustin commented:
Apollon, please ping when this is ready for review, I'll be happy to.

apoikos commented:
> I am not an expert on porting code to Python3, and I lack the resources to test this code.
> 
> For what it is worth, I attest that I have read the whole PR and I confirm that it looks like a good faith effort to update the semantics of a large codebase to Python 3.

Thanks a lot for your time and for going through the PR @dannyman!

apoikos commented:
@iustin, thanks, please go ahead!

I generated a new docker image (ganeti/ci:buster-py3) with the Python 3 dependencies, which means the CI is to be trusted. You could also use the same image if you want to run something locally and save some time.

I know that parts of this PR are not as pretty as they could, but I chose to trade some elegance for time. This would have been a good chance to do some large-scale refactoring, however I think we have neither the resources, nor the time to do so right now.

If this gets merged, I plan to release 3.0~alpha1 shortly after to start building packages and get some real-world testing if possible.

apoikos commented:
@iustin I pushed a couple more commits, including a couple of fixups. I didn't want to rebase right now in order not to interfere with your review. I'll rebase & squash when we're done.

The good news is, I was able to do a full vcluster QA run with the tests we run in the Debian package. It uncovered only a couple of minor issues - most of them in the QA suite itself - which I fixed.

iustin commented:
On 2019-12-09 14:04:43, Apollon Oikonomopoulos wrote:
> @iustin I pushed a couple more commits, including a couple of fixups. I didn't want to rebase right now in order not to interfere with your review. I'll rebase & squash when we're done.

Thank you. It's honestly a pain to review as it is, at 100 commits, so
much appreciated for not making it even more difficult :/

iustin

iustin commented on test/hs/Test/Ganeti/TestCommon.hs:
```diff
@@ -435,12 +441,26 @@ genIp6Addr = do
+      chunksOf _ [] = []
+      chunksOf n lst = take n lst : (chunksOf n $ drop n lst)
+      lPadTrim :: Int -> a -> [a] -> [a]
+      lPadTrim n p lst
```
Please add a comment what this does/example of how output would look. For tired eyes, it is not obvious.

iustin commented on qa/qa_utils.py:
```diff
@@ -347,7 +347,7 @@ def StartLocalCommand(cmd, _nolog_opts=False, log_cmd=True, **kwargs):
     print("%s %s" % (colors.colorize("Command:", colors.CYAN),
                      utils.ShellQuoteArgs(pcmd)))
-  return subprocess.Popen(cmd, shell=False, **kwargs)
+  return subprocess.Popen(cmd, shell=False, text="utf-8", **kwargs)
```
This and below confuse me greatly. The documentatio for Popen() argument states that the `text` argument:

- is an alias for universal_newlines, which is a boolean
- and was added in 3.7 only (we're still targetting 3.5 or 3.6?)

Did you mean instead `encoding` here?

apoikos replied:
Yes, I meant encoding, but misread the documentation. This worked because "utf-8" evaluates to True, but I'll fix it.

apoikos commented on lib/utils/io.py:
```diff
@@ -1053,3 +1055,19 @@ def CanRead(username, filename):
+  _tty = io.FileIO(device, "r+")
+  _buffered_tty = io.BufferedRWPair(_tty, _tty)
+
+  return io.TextIOWrapper(_buffered_tty)
```
Yes, that's a good idea. I resorted to calling flush() by hand, but line buffering should also work (although not for prompts that do not include a newline).

apoikos commented on test/py/ganeti.server.rapi_unittest.py:
```diff
@@ -147,8 +147,8 @@ def testIncompleteBasicAuth(self):
   def testInvalidBasicAuth(self):
-    for auth in ["!invalid=base!64.", base64.b64encode(" "),
-                 base64.b64encode("missingcolonchar")]:
+    for auth in ["!invalid=base!64.", base64.b64encode(b" ").decode("ascii"),
```
Ack, will add.

apoikos replied:
Actually I already implemented that for `ganeti.http_unittest.py`, so I'll just move it over to testutils.

apoikos commented on lib/utils/process.py:
```diff
@@ -558,7 +559,8 @@ def _RunCmdPipe(cmd, env, via_shell, cwd, interactive, timeout, noclose_fds,
                            env=env,
-                           cwd=cwd)
+                           cwd=cwd,
+                           encoding="utf-8")
```
Yes, I'll verify the versions once more and adjust configure.ac accordingly.

apoikos commented on src/Ganeti/PyValue.hs:
```diff
@@ -78,6 +82,24 @@ instance PyValue Char where
+--  * escapes double quotes
+--  * leaves all printable ASCII characters intact
+--  * encodes all other characters in \xYZ form
+instance PyValue ByteString where
```
Time to get my feet wet with Haskell unit testing I guess :)

apoikos replied:
By the way, you were right to point this out. QuickCheck helped point out a flaw in the implementation, failing to escape backslashes properly.

apoikos commented on lib/utils/io.py:
```diff
@@ -112,7 +134,7 @@ def ReadFile(file_name, size=-1, preread=None):
 
   """
-  f = open(file_name, "r")
+  f = open(file_name, "r", encoding="utf-8")
```
Me neither, but I don't think it warrants a constant. As you said, I don't think we'll need to switch encodings anytime soon (famous last words).

apoikos commented on lib/cmdlib/instance_create.py:
```diff
@@ -675,29 +675,29 @@ def _RevertToDefaults(self, cluster):
     nic_defs = cluster.SimpleFillNIC({})
     for nic in self.op.nics:
-      for name in constants.NICS_PARAMETERS:
+      for name in list(constants.NICS_PARAMETERS):
```
Ah, that's just oversight. I'll fix it.

apoikos commented on lib/utils/io.py:
```diff
@@ -230,17 +230,15 @@ def WriteFile(file_name, fn=None, data=None,
+          open_mode = "wb"
+          encoding = None
+        with open(fd, open_mode, closefd=False, encoding=encoding) as out:
+          out.write(data)
```
Yes, it does. open() returns a subclass of BufferedIOBase. For text mode, it uses line buffering by default. Now that I read the original code again, it's obvious that we wanted to do zero-copy writes, which is the opposite from using buffered/batched writes. Do you think it still makes sense? I think that switching this over to block buffering with e.g. an 8kB-long buffer shouldn't do much harm.

iustin replied:
I took a quick look at callers of WriteFile. Most of this data is small, so trying to do no-buffering doesn't matter much.

Also, this was doing no-buffering user-side, kernel side it was still a normal write going through the page-cache and not an `O_DIRECT` write, so I think this was rather trying to do something else - avoid partial writes caused by signals. That seems to me however a rare occurence, but who knows (how things behaved >10 years ago). In any case, [BufferedIOBase](https://docs.python.org/3.7/library/io.html#io.BufferedIOBase.write) says whole input is written, always, or an exception is raised, so this should go away indeed.

Note that below we do an os.fsync(fd) - I think we need first to call `fd.sync()` here to ensure the (Python/userspace) buffers are sent to the kernel before attempting to flush the kernel buffer.

apoikos replied:
> Note that below we do an os.fsync(fd) - I think we need first to call `fd.sync()` here to ensure the (Python/userspace) buffers are sent to the kernel before attempting to flush the kernel buffer.

I think that's not necessary, because BufferedIOReader flushes the buffer on close, and we're using it as a context manager which calls close() when done.

iustin replied:
Ah, right. To be honest the fd-vs-out distiction here is a bit confusing; but now the `closefd=False` makes sense :)

apoikos commented:
The PyValue test failures are due to the tests running with `LC_ALL=POSIX`, causing Haskell to decode Python's stdout as plain ASCII. I'll fix this later today.

apoikos commented:
@iustin: Thanks a lot for the review! I think I addressed all issues, interdiff can be found [here](https://github.com/ganeti/ganeti/compare/c1b1c669e956dcabb0aa43ca54800ae51672a078..apoikos:python3-next). I thought GH would be able to keep track of conversations across force-pushes, but I guess the comments were attached to particular commits. If I'd known that, I wouldn't have force-pushed now, sorry!

iustin commented:
> @iustin: Thanks a lot for the review! I think I addressed all issues, interdiff can be found [here](https://github.com/ganeti/ganeti/compare/c1b1c669e956dcabb0aa43ca54800ae51672a078..apoikos:python3-next). 

Thanks, interdiff looks fine. I would have some comments on the PyValue unittest (why is it fixed size rather than proper QC generator), but I don't have the energy for it, maybe later.

> I thought GH would be able to keep track of conversations across force-pushes, but I guess the comments were attached to particular commits rather than files + lines. If I'd known that, I wouldn't have force-pushed now, sorry!

Hmm, it is not a problem for this patch series, but then I wonder what is the proper way to address commets with GH, just add fixups? Or force push and ignore lost comments? Strange way to do reviews…

I'll approve this, please merge and let's move on trying to forget Py 2 past :)

apoikos commented:
@iustin, thanks again for your time. I know it was quite an amount to review, so your help is much appreciated!
Re: GH pull requests, I think the optimal workflow is to push incremental changes/fixups until everything is approved, and then rebase/squash before merging.
